include/dlaf/profiler.h

```diff
@@ -0,0 +1,133 @@
+//
+// Distributed Linear Algebra with Future (DLAF)
+//
+// Copyright (c) 2018-2019, ETH Zurich
+// All rights reserved.
+//
+// Please, refer to the LICENSE file in the root directory.
+// SPDX-License-Identifier: BSD-3-Clause
+//
+
+#pragma once
+
+#include <fstream>
+#include <string>
+#include <thread>
+#include <chrono>
+#include <map>
+#include <deque>
+
+namespace dlaf {
+namespace profiler {
+namespace details {
+
+  /// Data structure for storing a single profile entry
+  class task_data_t {
+    using clock_t = std::chrono::steady_clock;
+
+    std::string name_;
+    std::string group_;
+
+    /// data structure for the time information
+    struct {
+      std::thread::id id_;
+      std::size_t time_;
+    } start_data_, end_data_;
+
+    public:
+    void enter(const std::string& name, const std::string& group) {
+      name_ = name;
+      group_ = group;
+      start_data_ = { std::this_thread::get_id(), get_time() };
+    }
+
+    void leave() {
+      end_data_ = { std::this_thread::get_id(), get_time() };
+    }
+
+    /// @return value is valid just if the same thread has called enter and leave methods
+    std::size_t get_time() {
+      return std::chrono::duration_cast<std::chrono::nanoseconds>(
+            clock_t::now().time_since_epoch()
+          ).count();
+    };
+
+    friend std::ostream& operator<<(std::ostream& os, const task_data_t& task_data) {
+      os << task_data.name_ << ", ";
+      os << task_data.group_ << ", ";
+      os << task_data.start_data_.id_ << ", ";
+      os << task_data.start_data_.time_ << ", ";
+      os << task_data.end_data_.id_ << ", ";
+      os << task_data.end_data_.time_;
+      return os;
+    }
+  };
+
+}
+
+/// Global profiler manager
+///
+/// On destruction it dumps on a csv file all tasks
+class Manager {
+  public:
+  static Manager& get_global_profiler() {
+    static Manager global_;
+    return global_;
+  }
+
+  ~Manager() {
+    std::ofstream profiler_report("report.csv");  // TODO adapt for multiple nodes
```

Comment:
The profile name should be `<name>_<rank_id>_<num_ranks>.csv`
What about giving the possibility to set the profile name `<name>`?

Reply:
I agree with you, moreover passing the filename as a parameter would keep MPI away from the profiler.